include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -1160,6 +1293,141 @@ namespace VectorTools
 
 
 
+  template <int components, int fe_degree, int dim, typename VectorType, int spacedim>
+  void project_generic (const Mapping<dim, spacedim>                              &mapping,
+                        const DoFHandler<dim, spacedim>                           &dof,
+                        const ConstraintMatrix                                    &constraints,
+                        const Quadrature<dim>                                     &quadrature,
+                        const Function<spacedim, typename VectorType::value_type> &function,
+                        VectorType                                                &vec_result,
+                        const bool                                                 enforce_zero_boundary,
+                        const Quadrature<dim-1>                                   &q_boundary,
+                        const bool                                                 project_to_boundary_first)
+  {
+    const parallel::Triangulation<dim,spacedim> *parallel_tria =
+      dynamic_cast<const parallel::Triangulation<dim,spacedim>*>(&dof.get_tria());
+    Assert (project_to_boundary_first == false, ExcNotImplemented());
+    Assert (enforce_zero_boundary == false, ExcNotImplemented());
+    (void) enforce_zero_boundary;
+    (void) project_to_boundary_first;
+    (void) q_boundary;
+
+    const IndexSet locally_owned_dofs = dof.locally_owned_dofs();
+    IndexSet locally_relevant_dofs;
+    DoFTools::extract_locally_relevant_dofs(dof, locally_relevant_dofs);
+
+    typedef typename VectorType::value_type number;
+    Assert (dof.get_fe().n_components() == function.n_components,
+            ExcDimensionMismatch(dof.get_fe().n_components(),
+                                 function.n_components));
+    Assert (vec_result.size() == dof.n_dofs(),
+            ExcDimensionMismatch (vec_result.size(), dof.n_dofs()));
+    Assert (dof.get_fe().degree == fe_degree,
+            ExcDimensionMismatch(fe_degree, dof.get_fe().degree));
+    Assert (dof.get_fe().n_components() == components,
+            ExcDimensionMismatch(components, dof.get_fe().n_components()));
+
+    // set up mass matrix and right hand side
+    typename MatrixFree<dim,number>::AdditionalData additional_data;
+    additional_data.tasks_parallel_scheme =
+      MatrixFree<dim,number>::AdditionalData::partition_color;
+    if (parallel_tria)
+      additional_data.mpi_communicator = parallel_tria->get_communicator();
```

Comment:
The other question is whether it actually makes sense to be able to specify a separate communicator here. Given a mesh, the algorithm has no other choice than using the communicator of the mesh because it won't be able to get different ownership or locally owned ranges anyway... What do you @masterleinad @davydden think?